pypeit/reduce.py

```diff
@@ -0,0 +1,719 @@
+
+        # Setup the parameters sets for this object. NOTE: This uses objtype, not frametype!
+        self.objtype = objtype
+        self.par = pypeitpar.ScienceImagePar() if par is None else par
+        self.frame_par = pypeitpar.FrameGroupPar(objtype) if frame_par is None else frame_par
+        self.proc_par = self.frame_par['process']
+
+
+        # Instantiation attributes for this object
+        self.spectrograph = spectrograph
+        self.tslits_dict = tslits_dict
+        self.mask = mask
+        self.slitmask = pixels.tslits2mask(self.tslits_dict)
+        # Now add the slitmask to the mask (i.e. post CR rejection in proc)
+        self.mask = processimages.ProcessImages.update_mask_slitmask(self.mask, self.slitmask)
+
+        self.ir_redux = ir_redux
+        self.det = det
+        self.binning = binning
+        self.setup = setup
+        self.pypeline = spectrograph.pypeline
+
+        self.steps = []
+
+        # Other attributes that will be set later during object finding,
+        # sky-subtraction, and extraction
+        self.tilts = None  # used by extract
+        self.waveimage = None  # used by extract
+        self.maskslits = None  # used in find_object and extract
+
+        # Key outputs images for extraction
+        self.sciimg = None
+        self.sciivar = None
+        self.ivarmodel = None
+        self.objimage = None
+        self.skyimage = None
+        self.global_sky = None
+        self.skymask = None
+        self.outmask = None
+        self.extractmask = None
+        # SpecObjs object
+        self.sobjs_obj = None  # Only object finding but no extraction
+        self.sobjs = None  # Final extracted object list with trace corrections applied
+
+        # Other bookeeping internals
+        self.crmask = None
+
+
+    def _chk_objs(self, items):
+        """
+
+        Args:
+            items:
+
+        Returns:
+
+        """
+        for obj in items:
+            if getattr(self, obj) is None:
+                msgs.warn('You need to generate {:s} prior to this step..'.format(obj))
+                if obj in ['sciimg', 'sciivar', 'rn2_img']:
+                    msgs.warn('Run the process() method')
+                elif obj in ['sobjs_obj']:
+                    msgs.warn('Run the find_objects() method')
+                elif obj in['global_sky']:
+                    msgs.warn('Run the global_skysub() method')
+                elif obj in ['tilts', 'tslits_dict'] :
+                    msgs.warn('Calibrations missing: these were required to run find_objects() '
+                              'and global_skysub()')
+                elif obj in ['waveimg']:
+                    msgs.warn('Calibrations missing: waveimg must be input as a parameter. Try '
+                              'running calibrations')
+                return False
+        return True
+
+
+    def find_objects(self, image, ivar, std=False, ir_redux=False, std_trace=None, maskslits=None,
```

Comment:
This is a rather odd wrapper.  Is it only for the `ir_redux` if/else?

Reply:
This wrapper is necessary. Look more closely at the code. The find objects needs to be able to work both for IR and not IR, and for IR it must find negative objects.